File: web/src/api/calendarCacheStore.ts
import type { CalendarItem } from "@/api/types"

/** Soft TTL before returning to Calendar forces a background revalidate. */
export const CALENDAR_CACHE_SOFT_TTL_MS = 5 * 60 * 1000

const STORAGE_PREFIX = "family-carpool.calendar-cache:"

export type CalendarCacheSnapshot = {
  adultId: string
  circleId: string
  from: string
  to: string
  items: CalendarItem[]
  fetchedAt: number
}

function storageKey(adultId: string, circleId: string): string {
  return `${STORAGE_PREFIX}${adultId}:${circleId}`
}

/**
 * Persists the last successful calendar GET per (adultId, circleId).
 * Not a secret store — calendar enrichment is adult-scoped but not credentials.
 */
export class CalendarCacheStore {
  private readonly storage: Pick<
    Storage,
    "getItem" | "setItem" | "removeItem" | "key" | "length"
  >

  constructor(
    storage: Pick<
      Storage,
      "getItem" | "setItem" | "removeItem" | "key" | "length"
    > = localStorage,
  ) {
    this.storage = storage
  }

  load(adultId: string, circleId: string): CalendarCacheSnapshot | null {
    const raw = this.storage.getItem(storageKey(adultId, circleId))
    if (!raw) {
      return null
    }
    try {
      const parsed = JSON.parse(raw) as CalendarCacheSnapshot
      if (
        parsed.adultId !== adultId ||
        parsed.circleId !== circleId ||
        typeof parsed.from !== "string" ||
        typeof parsed.to !== "string" ||
        !Array.isArray(parsed.items) ||
        typeof parsed.fetchedAt !== "number"
      ) {
        return null
      }
      return parsed
    } catch {
      return null
    }
  }

  save(snapshot: CalendarCacheSnapshot): void {
    this.storage.setItem(
      storageKey(snapshot.adultId, snapshot.circleId),
      JSON.stringify(snapshot),
    )
  }

  /** Replace one calendar row in the persisted snapshot when present. */
  patchItem(adultId: string, circleId: string, updated: CalendarItem): void {
    const existing = this.load(adultId, circleId)
    if (!existing) {
      return
    }
    this.save({
      ...existing,
      items: existing.items.map((row) =>
        row.source === updated.source && row.id === updated.id ? updated : row,
      ),
    })
  }

  clear(adultId: string, circleId: string): void {
    this.storage.removeItem(storageKey(adultId, circleId))
  }

  /** Remove every calendar snapshot (sign-out). */
  clearAll(): void {
    const keys: string[] = []
    for (let i = 0; i < this.storage.length; i++) {
      const key = this.storage.key(i)
      if (key?.startsWith(STORAGE_PREFIX)) {
        keys.push(key)
      }
    }
    for (const key of keys) {
      this.storage.removeItem(key)
    }
  }

  isStale(snapshot: Pick<CalendarCacheSnapshot, "fetchedAt">, nowMs: number = Date.now()): boolean {
    return nowMs - snapshot.fetchedAt > CALENDAR_CACHE_SOFT_TTL_MS
  }
}

/** Later of two ISO-8601 UTC instants (lexicographic compare is safe for this format). */
export function maxIsoInstant(a: string, b: string): string {
  return a >= b ? a : b
}

File: mobile/sharedLogic/src/commonMain/kotlin/org/example/project/CalendarCacheStore.kt
package org.example.project

import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** Soft TTL before returning to Calendar forces a background revalidate. */
const val CALENDAR_CACHE_SOFT_TTL_MS: Long = 5 * 60 * 1000L

private const val STORAGE_PREFIX = "family-carpool.calendar-cache:"

@Serializable
data class CalendarCacheSnapshot(
    val adultId: String,
    val circleId: String,
    val from: String,
    val to: String,
    val items: List<CalendarItem>,
    val fetchedAt: Long,
)

/**
 * Persists the last successful calendar GET per (adultId, circleId).
 * Not a secret store — do not use encrypted token storage.
 */
interface CalendarCacheStore {
    fun load(
        adultId: String,
        circleId: String,
    ): CalendarCacheSnapshot?

    fun save(snapshot: CalendarCacheSnapshot)

    fun patchItem(
        adultId: String,
        circleId: String,
        updated: CalendarItem,
    )

    fun clear(
        adultId: String,
        circleId: String,
    )

    fun clearAll()

    fun isStale(
        fetchedAt: Long,
        nowMs: Long,
    ): Boolean = nowMs - fetchedAt > CALENDAR_CACHE_SOFT_TTL_MS
}

/** Later of two ISO-8601 UTC instants (lexicographic compare is safe for this format). */
fun maxIsoInstant(
    a: String,
    b: String,
): String = if (a >= b) a else b

/** Platform-agnostic string KV used by [JsonCalendarCacheStore]. */
interface CalendarCacheKeyValueStore {
    fun getString(key: String): String?

    fun putString(
        key: String,
        value: String,
    )

    fun remove(key: String)

    fun keysWithPrefix(prefix: String): List<String>
}

class JsonCalendarCacheStore(
    private val kv: CalendarCacheKeyValueStore,
    private val json: Json =
        Json {
            ignoreUnknownKeys = true
            encodeDefaults = true
        },
) : CalendarCacheStore {
    override fun load(
        adultId: String,
        circleId: String,
    ): CalendarCacheSnapshot? {
        val raw = kv.getString(storageKey(adultId, circleId)) ?: return null
        return runCatching {
            val parsed = json.decodeFromString<CalendarCacheSnapshot>(raw)
            if (parsed.adultId != adultId || parsed.circleId != circleId) {
                null
            } else {
                parsed
            }
        }.getOrNull()
    }

    override fun save(snapshot: CalendarCacheSnapshot) {
        kv.putString(
            storageKey(snapshot.adultId, snapshot.circleId),
            json.encodeToString(snapshot),
        )
    }

    override fun patchItem(
        adultId: String,
        circleId: String,
        updated: CalendarItem,
    ) {
        val existing = load(adultId, circleId) ?: return
        save(
            existing.copy(
                items =
                    existing.items.map { row ->
                        if (row.source == updated.source && row.id == updated.id) updated else row
                    },
            ),
        )
    }

    override fun clear(
        adultId: String,
        circleId: String,
    ) {
        kv.remove(storageKey(adultId, circleId))
    }

    override fun clearAll() {
        for (key in kv.keysWithPrefix(STORAGE_PREFIX)) {
            kv.remove(key)
        }
    }

    private fun storageKey(
        adultId: String,
        circleId: String,
    ): String = "$STORAGE_PREFIX$adultId:$circleId"
}

class InMemoryCalendarCacheKeyValueStore : CalendarCacheKeyValueStore {
    private val map = linkedMapOf<String, String>()

    override fun getString(key: String): String? = map[key]

    override fun putString(
        key: String,
        value: String,
    ) {
        map[key] = value
    }

    override fun remove(key: String) {
        map.remove(key)
    }

    override fun keysWithPrefix(prefix: String): List<String> =
        map.keys.filter { it.startsWith(prefix) }
}

class InMemoryCalendarCacheStore :
    CalendarCacheStore by JsonCalendarCacheStore(InMemoryCalendarCacheKeyValueStore())
